common/nym-lp/src/codec.rs

```diff
@@ -3,105 +3,382 @@
 
 use crate::LpError;
 use crate::message::{
-    ClientHelloData, EncryptedDataPayload, HandshakeData, KKTRequestData, KKTResponseData,
-    LpMessage, MessageType,
+    ClientHelloData, EncryptedDataPayload, ForwardPacketData, HandshakeData, KKTRequestData,
+    KKTResponseData, LpMessage, MessageType, SubsessionKK1Data, SubsessionKK2Data,
+    SubsessionReadyData,
 };
-use crate::packet::{LpHeader, LpPacket, TRAILER_LEN};
-use bytes::BytesMut;
+use crate::packet::{LpHeader, LpPacket, OuterHeader, TRAILER_LEN};
+use bytes::{BufMut, BytesMut};
 
-/// Parses a complete Lewes Protocol packet from a byte slice (e.g., a UDP datagram payload).
+/// Size of outer header (receiver_idx + counter) - always cleartext
+pub const OUTER_HEADER_SIZE: usize = OuterHeader::SIZE; // 12 bytes
+
+/// Size of inner prefix (proto + reserved) - cleartext or encrypted depending on mode
+const INNER_PREFIX_SIZE: usize = 4; // proto(1) + reserved(3)
+use chacha20poly1305::{
+    aead::{AeadInPlace, KeyInit},
+    ChaCha20Poly1305, Key, Nonce, Tag,
+};
+use zeroize::{Zeroize, ZeroizeOnDrop};
+
+/// Outer AEAD key for LP packet encryption.
 ///
-/// Assumes the input `src` contains exactly one complete packet. It does not handle
-/// stream fragmentation or provide replay protection checks (these belong at the session level).
-pub fn parse_lp_packet(src: &[u8]) -> Result<LpPacket, LpError> {
-    // Minimum size check: LpHeader + Type + Trailer (for 0-payload message)
-    let min_size = LpHeader::SIZE + 2 + TRAILER_LEN;
-    if src.len() < min_size {
-        return Err(LpError::InsufficientBufferSize);
-    }
+/// Derived from PSK using Blake3 KDF with domain separation.
+/// Used for opportunistic encryption: before PSK packets are cleartext,
+/// after PSK packets have encrypted payload and authenticated header.
+///
+/// # Security: Nonce Reuse Prevention
+///
+/// ChaCha20-Poly1305 requires unique nonces per key. The counter starts at 0
+/// for each session, which is safe because:
+///
+/// 1. **PSK is always fresh**: Each handshake uses PSQ 
+///    with a client-generated random salt. This ensures a unique
+///    PSK for every session, even between the same client-gateway pair.
+///
+/// 2. **Key derivation**: `outer_key = Blake3_KDF("lp-outer-aead", PSK)`.
+///    Different PSK → different outer_key → nonce reuse impossible.
+///
+/// 3. **No PSK persistence**: PSK handles are not stored/reused across sessions.
+///    Each connection performs fresh KKT+PSQ handshake.
+///
+#[derive(Clone, Zeroize, ZeroizeOnDrop)]
+pub struct OuterAeadKey {
+    key: [u8; 32],
```

Comment:
would it make sense to just explicitly hold here an instance of `chacha20poly1305::Key`?